<!-- Robor utilities -->
<launch>  

  <arg name="simulator"/>
  <arg name="manager"/>
  <arg name="loop_rate"/>

  <arg name="use_only_odometry"/>
  <arg name="use_optitrack"/>
  <arg name="use_map"/>
  <arg name="do_mapping"/>

  <arg name="map_frame"/>
  <arg name="odom_frame"/>
  <arg name="robot_frame"/>
  <arg name="reference_frame"/>

  <arg name="front_scanner_frame"/>
  <arg name="rear_scanner_frame"/>
  <arg name="UAM_scanner_frame"/>

  <group if="$(arg simulator)">
    <node pkg="nodelet" type="nodelet" name="simulator" args="load robor_utilities/Simulator $(arg manager)" machine="usr_pc">
      <param name="active"           value="true"/>
      <param name="continuous_angle" value="true"/>

      <param name="loop_rate"        value="$(arg loop_rate)"/>
      <param name="time_constant"    value="0.0"/>
      <param name="time_delay"       value="0.0"/>

      <param name="init_x"           value="1.0"/>
      <param name="init_y"           value="0.7"/>
      <param name="init_theta"       value="1.2"/>

      <param name="parent_frame_id"     value="$(arg odom_frame)"/>
      <param name="child_frame_id"      value="$(arg robot_frame)"/>

      <remap from="simulator_state"     to="robot_state"/>
    </node>

    <node pkg="rviz" type="rviz" name="rviz" args="-d $(find robor_utilities)/resources/simulator.rviz" machine="usr_pc" required="true"/>
  </group>

  <group unless="$(arg simulator)">
    <node pkg="nodelet" type="nodelet" name="state_estimator" args="load robor_utilities/StateEstimator $(arg manager)" machine="yb_pc" unless="$(arg use_only_odometry)">
      <param name="active"           value="true"/>

      <param name="loop_rate"        value="$(arg loop_rate)"/>

      <param name="angular_var"      value="0.01"/>
      <param name="angular_rate_var" value="0.01"/>
      <param name="linear_var"       value="0.01"/>
      <param name="linear_rate_var"  value="0.01"/>

      <param name="parent_frame_id"  value="$(arg odom_frame)"/>
      <param name="child_frame_id"   value="$(arg robot_frame)"/>
    </node>

    <node pkg="rviz" type="rviz" name="rviz" args="-d $(find robor_utilities)/resources/robor.rviz" machine="usr_pc" required="true" unless="$(arg do_mapping)"/>
    <node pkg="rviz" type="rviz" name="rviz" args="-d $(find robor_utilities)/resources/mapping.rviz" machine="usr_pc" required="true" if="$(arg do_mapping)"/>
  </group>

  <node pkg="nodelet" type="nodelet" name="reference_generator" args="load robor_utilities/ReferenceGenerator $(arg manager)" machine="yb_pc">
    <param name="active"                 value="true"/>
    <param name="loop_rate"              value="$(arg loop_rate)"/>

    <param name="parent_frame_id"        value="$(arg odom_frame)"/>
    <param name="child_frame_id"         value="$(arg reference_frame)"/>

    <param name="continuous_angle"       value="false"/>    
    <param name="trajectory_paused"      value="false"/>
    <param name="trajectory_stopped"     value="false"/>

    <param name="trajectory_type"        value="0"/>

    <param name="initial_x"              value="0.0"/>
    <param name="initial_y"              value="0.0"/>
    <param name="initial_theta"          value="0.0"/>
    <param name="linear_velocity"        value="0.1"/>
    <param name="harmonic_period"        value="30.0"/>
    <param name="harmonic_radius_x"      value="1.0"/>
    <param name="harmonic_radius_y"      value="0.5"/>
    <param name="harmonic_multiplier_x"  value="1.0"/>
    <param name="harmonic_multiplier_y"  value="2.0"/>
  </node>

  <node pkg="nodelet" type="nodelet" name="data_recorder" args="load robor_utilities/DataRecorder $(arg manager)" machine="yb_pc">
    <param name="active"                    value="false"/>
    <param name="recording"                 value="false"/>
    <param name="loop_rate"                 value="$(arg loop_rate)"/>

    <param name="record_pose"               value="true"/>
    <param name="record_reference_pose"     value="true"/>
    <param name="record_velocity"           value="true"/>
    <param name="record_reference_velocity" value="true"/>
    <param name="record_controls"           value="true"/>
  </node>

  <node pkg="robor_utilities" type="youbot_visualizer_node" name="youbot_visualizer" machine="usr_pc">
    <param name="active"         value="false"/>
    <param name="emit_yb_marker" value="false"/>
    <param name="loop_rate"      value="30.0"/>
    <param name="frame_id"       value="$(arg robot_frame)"/>
  </node>

  <!--
  <node name="joy" pkg="joy" type="joy_node" machine="usr_pc">
    <param name="dev" type="string" value="/dev/input/js0"/>
    <param name="deadzone"          value="0.08"/>
    <param name="autorepeat_rate"   value="0"/>
  </node>
  -->

</launch>
<!-- -->
